Redesign tab bar

pinussilvestrus opened the pull request:
Closes #2563 

**What's inside**

* add `getIcon` handler to tab providers
* redesign tab bar (cf. https://github.com/camunda/camunda-modeler/issues/2563)
* updated unsaved state

![image](https://user-images.githubusercontent.com/9433996/144033996-7bfdb3f4-a244-4e86-afd6-c5ec9a4c5652.png)

![image](https://user-images.githubusercontent.com/9433996/144115018-1cac52f9-0dd0-423f-a765-2a0f50086e59.png)

![image](https://user-images.githubusercontent.com/9433996/144114849-e369719d-91e6-45c1-b003-811a3db566ff.png)



Find artifacts here to check it out (cf. [Build](https://github.com/camunda/camunda-modeler/actions/runs/1506683299))

* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-linux-x64.tar.gz
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-mac.dmg
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-mac.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-win-ia32.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-win-x64.zip

pinussilvestrus commented:
I decided to keep the `TabLinks` inside the `primitives` to avoid bigger re-arrangements when not needed (although the TabLinks component is less primitive than before).

barmac commented:
Found a glitch for a long name: 
![image](https://user-images.githubusercontent.com/28307541/143558199-d698fefd-b145-4e66-a2ad-919e9084522d.png)

barmac commented:
Is this part of the design to display the dirty marker only for the active tab?

pinussilvestrus commented:
> The "X" button is overflowing so it looks like an arrow. When I click the dirty marker, it triggers the close action.

Seems like I broke the existing dirty marker styles with this one. That shall be fixed, thanks!

> Is this part of the design to display the dirty marker only for the active tab?

The way the dirty marker is currently implemented is heavily dependant on the close handle. Since the close handle is not present on non-active tabs, the dirty marker also won't show up. We will update the whole dirty marker topic in a follow-up, that's why I completely left it untouched.

barmac commented:
@andreasgeier The contrast here is super small:
![image](https://user-images.githubusercontent.com/28307541/143562926-fa7d5779-8453-42e9-a79d-1aa2d22e9766.png)

barmac reviewed (changes requested):
Generally, it looks good. There are a few issues to address before we merge this.

pinussilvestrus commented on client/src/app/primitives/Tabbed.less:
```diff
@@ -17,176 +16,164 @@
+            display: inline;
+          }
+          
+          .tab_icon-dirty {
```
That actually causes the bug described in https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-979818838 👍

andreasgeier commented:
@pinussilvestrus 

I adjusted the not-active state to consider the feedback from @barmac regarding contrast. See updated [prototype v16](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v16/) for the new colors. It's the same color like the active state, I limited the differentiation to the underline for active state.

<img width="546" alt="Screen Shot 2021-11-26 at 18 02 32" src="https://user-images.githubusercontent.com/938244/143613016-e3d8a24f-db0a-4bef-a21c-0fb4bf6243e3.png">

Also, please correct the style of the active state line. I made a CSS mistake in the prototype and the line got truncated by other lines. Again, see [prototype v16](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v16/) for the fix. This is the expected style (2px height, overlap other lines):

<img width="688" alt="Screen Shot 2021-11-26 at 18 01 47" src="https://user-images.githubusercontent.com/938244/143612961-0017c34b-08b6-456d-91df-21030862f336.png">

In the meantime, I added the unsaved marker as well. See [tab bar design issue](https://github.com/camunda/camunda-modeler-design/issues/37) and [prototype v16](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v16/).

pinussilvestrus commented:
Markes as [WIP] to add requested changes
* add the new dirty marker
* adjust active and hover bottom line (2px height, overlay other lines)
* react to tiny state

![image](https://user-images.githubusercontent.com/9433996/143842217-b7a608a7-e5a2-49aa-86db-4040423393a9.png)

pinussilvestrus commented:
Regarding this one 

> This results in the tab line being properly set, but don't move when the window got resized (cf. screencast).

We can potentially re-render the lines once the window get resized (sketched via 077f5c4bef46c76b217f5aed3e361905b5f6815d).

![Kapture 2021-11-30 at 15 57 37](https://user-images.githubusercontent.com/9433996/144071037-122f8273-19cc-4c2d-b168-951cfcd2bf4d.gif)

Let's discuss this one in an orderly manner later this week @barmac @andreasgeier.

pinussilvestrus commented:
Thanks to @andreasgeier we were able to resolve the tab bottom line with CSS only, no fixed positioning is needed 👍 

This is up for review now, the artifacts are up-to-date.

nikku reviewed (commented):
It seems we accidentally broke re-arranging the tabs. Can you have a look?

As you can see in the capture below it does work, sometimes, only.

![capture XozHoQ_optimized](https://user-images.githubusercontent.com/58601/144233122-f14cd87a-1657-454f-9ca9-74c1699103fb.gif)

nikku commented:
Other than that, works very nice!

pinussilvestrus commented:
Thanks for the report for drag & drop! That should be resolved now, artefacts are updated.

barmac commented:
I am in the review right now (commit `a432241`).

barmac commented:
https://user-images.githubusercontent.com/28307541/144406925-3d7579d5-fedb-45ba-889c-26942b28bd4d.mov

I have to change the tab to trigger change from the small tab to normal tab appearance.

barmac commented on client/src/app/primitives/TabLinks.js:
```diff
@@ -147,57 +105,139 @@ export default class TabLinks extends PureComponent {
+
+  const tabRef = React.useRef(0);
+
+  // set tab to <small> once a defined threshold is reached (be responsive)
+  React.useEffect(() => {
+    if (!tabRef.current) {
+      return;
+    }
+
+    const tabNode = tabRef.current;
+    const width = tabNode.getBoundingClientRect().width;
+
+    tabNode.classList.toggle(TAB_SMALL_SELECTOR, width < SMALL_TAB_WIDTH);
+  });
```
Solution sketch for https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-984508557

```suggestion

  const tabNode = tabRef.current;

  // set tab to <small> once a defined threshold is reached (be responsive)
  React.useEffect(() => {
    if (!tabNode) {
      return;
    }

    const resizeObserver = new ResizeObserver(() => {
      const width = tabNode.getBoundingClientRect().width;

      tabNode.classList.toggle(TAB_SMALL_SELECTOR, width < SMALL_TAB_WIDTH);
    });

    resizeObserver.observe(tabNode);

    return () => resizeObserver.disconnect();
  }, [ tabNode ]);
```

barmac replied:
So instead of checking the width on render, we do this whenever node is resized.

pinussilvestrus replied:
I think was is missing here is to react to the _initial size_ of a tab (first render). E.g. when I open 15+ tabs at once, it wouldn't react because there is no resizing included.

pinussilvestrus replied:
Okay, actually the initial state would be respected as well. Instead, I'd probably handle the class via state, as we do with `dirty` and `active`. The reason is that it would override the classes on every render. Something like

```js
const [ small, setSmall ] = React.useState(false);

const tabNode = tabRef.current;

React.useEffect(() => {
  if (!tabNode) {
    return;
  }

  const resizeObserver = new ResizeObserver(() => {
    const width = tabNode.getBoundingClientRect().width;
    setSmall(width < SMALL_TAB_WIDTH);
  });

  resizeObserver.observe(tabNode);
  return () => resizeObserver.disconnect();
}, [ tabNode ]);

return (
  <div
    className={ classNames('tab', {
      'tab--active': active,
      'tab--dirty': dirty,
      'tab--small': small
      }) } >
      
      // ...
  </div>
); 
```

What do you think?

barmac replied:
LGTM

barmac commented:
> It seems we accidentally broke re-arranging the tabs. Can you have a look?

This works for me now :)

pinussilvestrus commented:
I pushed the resize observer behavior on top as suggested in https://github.com/camunda/camunda-modeler/pull/2589#discussion_r760983981. I don't build new artifacts for this but wait for design review first.

pinussilvestrus commented:
@andreasgeier will you review this one in particular or shall we wait once the tab action is added for an overall design review?

andreasgeier commented:
@pinussilvestrus I'm on it.

- entity icons are blurred because they are scaled. please ensure using a 100% display with sharp lines.
- font size should be 14px
- dirty marker should be 8x8px and should have a border in dark blue
- close icon looks scaled

please adjust the overall style to match the proto.

proto vs implementation:

![Screen Shot 2021-12-03 at 15 40 01](https://user-images.githubusercontent.com/938244/144620896-d0e0b4a5-cb64-40fb-b836-17b0300e3a3c.png)

andreasgeier commented:
The tab bar looks aery in the prototype because it is part of the unified header design. In the current implementation, it might look compressed because it is docked to the window frame. I think we should fix this by using the same height as the status bar (32px) but I can't judge it with the current styling issues.

We could try this out at a later state, I just wanted to mention it if you could easily try out and paste a screenshot.

pinussilvestrus commented:
Thanks @andreasgeier, I will follow up with the tiny fixes next week 👍

andreasgeier commented:
To distinguish between the different states, I would like to propose these title tags:

Newly created file, not yet saved: `(new file) – unsaved`
Existing file, untouched: `path/to/name.bpmn`
Existing file in dirty state: `path/to/name.bpmn – unsaved`

pinussilvestrus commented:
@andreasgeier I adjusted the design accordingly (unscaled icons, font-size 14px, dirty marker 8x8px). The titles are also updated to the one [you suggested](https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-985592073). Please note that there is currently bug related to unsaved --> then saved tabs, which is not related to this PR (cf. https://github.com/camunda/camunda-modeler/issues/2581). The artefacts are also updated.

I also increased the size of the tab bar to the same as the status bar. See the following screenshot:

<img width="1140" alt="Bildschirmfoto 2021-12-05 um 20 40 35" src="https://user-images.githubusercontent.com/9433996/144808338-a7278a92-0051-4943-8b28-db63f0a8b4af.png">

andreasgeier commented:
As discussed, please adjust the entity icon (and in relation the dirty marker) to be centered between divider and tab title.
Everything else looks as expected. Nice Work!

✅ &nbsp; Approved! Pls merge manually after the fix 🙂

pinussilvestrus commented:
Last styling issues are pushed, let's merge this one 👍
